test/extended/operators/workload_guarantees.go

```diff
@@ -0,0 +1,269 @@
+		}
+		// iterate over the references to find valid images
+		infra, err := oc.AdminConfigClient().ConfigV1().Infrastructures().Get(context.Background(), "cluster", metav1.GetOptions{})
+		if err != nil {
+			e2e.Failf("unable to get infrastructure: %v", err)
+		}
+		var inHAMode bool
+		if infra.Status.InfrastructureTopology == oapi.HighlyAvailableTopologyMode {
+			inHAMode = true
+		}
+
+		// workloads that are broken and need to be fixed, every entry here must have a bug associated
+		// e.g. "<mode>/<kind>/openshift-x/foo": "link to bug", <mode> can be HA or non-HA
+		knownBugs := map[string]string{
+			"HA/Deployment/openshift-image-registry/image-registry": "https://bugzilla.redhat.com/show_bug.cgi?id=1945387",
+			"HA/Deployment/openshift-ingress/router-default":        "https://bugzilla.redhat.com/show_bug.cgi?id=1945393",
+			"HA/Deployment/openshift-monitoring/prometheus-adapter": "https://bugzilla.redhat.com/show_bug.cgi?id=1948711",
+			"HA/Deployment/openshift-monitoring/thanos-querier":     "https://bugzilla.redhat.com/show_bug.cgi?id=1948711",
+			"HA/StatefulSet/openshift-monitoring/prometheus-k8s":    "https://bugzilla.redhat.com/show_bug.cgi?id=1949262",
+			"HA/StatefulSet/openshift-monitoring/alertmanager-main": "https://bugzilla.redhat.com/show_bug.cgi?id=1949262",
+		}
+
+		masterReq, err := labels.NewRequirement("node-role.kubernetes.io/master", selection.Exists, []string{})
+		if err != nil {
+			e2e.Failf("unable to build label requirement: %v", err)
+		}
+
+		var knownBroken []string
+		var invalidWorkloads []string
+		var replicas int32
+		for _, workload := range items {
+			var meta *metav1.ObjectMeta
+			var template *corev1.PodTemplateSpec
+			switch t := workload.(type) {
+			case *appsv1.Deployment:
+				t.Kind = "Deployment"
+				if t.Spec.Replicas == nil {
+					continue
+				}
+				meta = &t.ObjectMeta
+				template = &t.Spec.Template
+				replicas = *t.Spec.Replicas
+			case *appsv1.StatefulSet:
+				t.Kind = "StatefulSet"
+				if t.Spec.Replicas == nil {
+					continue
+				}
+				meta = &t.ObjectMeta
+				template = &t.Spec.Template
+				replicas = *t.Spec.Replicas
+			default:
+				panic("not an object")
+			}
+
+			if !strings.HasPrefix(meta.Namespace, "openshift-") {
+				continue
+			}
+
+			// exclude workloads that are master specific
+			if template.Spec.NodeSelector != nil {
+				if masterReq.Matches(labels.Set(template.Spec.NodeSelector)) {
+					continue
+				}
+			}
+			var toleratesMaster bool
+			for _, tolerates := range template.Spec.Tolerations {
+				if tolerates.Key == "node-role.kubernetes.io/master" && tolerates.Operator == corev1.TolerationOpExists {
+					toleratesMaster = true
+				}
+			}
+			if toleratesMaster {
+				continue
+			}
+
+			key := fmt.Sprintf("%s/%s/%s", workload.GetObjectKind().GroupVersionKind().Kind, meta.Namespace, meta.Name)
+
+			if inHAMode {
+				if replicas == 1 {
+					// TODO: We need to check if this is a misconfiguration.
```

Comment:
There are some components that are have just 1 replica today, not sure if upgrades will depend on the availability of them.